From 4556798c9f6658a0cc66371fb6599c9b074336a7 Mon Sep 17 00:00:00 2001
Message-Id: <4556798c9f6658a0cc66371fb6599c9b074336a7.1303219823.git.jdenemar@redhat.com>
From: Jiri Denemark <jdenemar@redhat.com>
Date: Fri, 25 Mar 2011 12:03:00 +0100
Subject: [PATCH] daemon: Avoid resetting errors before they are reported

CVE-2011-1486

Commit f44bfb7 was supposed to make sure no additional libvirt API (esp.
*Free) is called before remoteDispatchConnError() is called on error.
However, the patch missed two instances.
(cherry picked from commit 55cc591fc18e87b29febf78dc5b424b7c12f7349)
---
 daemon/remote.c |    6 ++++--
 1 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/daemon/remote.c b/daemon/remote.c
index 9e2dff5..caa9003 100644
--- a/daemon/remote.c
+++ b/daemon/remote.c
@@ -4511,12 +4511,13 @@ remoteDispatchStoragePoolListVolumes (struct qemud_server *server ATTRIBUTE_UNUS
     ret->names.names_len =
         virStoragePoolListVolumes (pool,
                                    ret->names.names_val, args->maxnames);
-    virStoragePoolFree(pool);
     if (ret->names.names_len == -1) {
         VIR_FREE(ret->names.names_val);
         remoteDispatchConnError(rerr, conn);
+        virStoragePoolFree(pool);
         return -1;
     }
+    virStoragePoolFree(pool);
 
     return 0;
 }
@@ -4540,11 +4541,12 @@ remoteDispatchStoragePoolNumOfVolumes (struct qemud_server *server ATTRIBUTE_UNU
     }
 
     ret->num = virStoragePoolNumOfVolumes (pool);
-    virStoragePoolFree(pool);
     if (ret->num == -1) {
         remoteDispatchConnError(rerr, conn);
+        virStoragePoolFree(pool);
         return -1;
     }
+    virStoragePoolFree(pool);
 
     return 0;
 }
-- 
1.7.5.rc1

